### PR TITLE
[WIP] Documentation improvement (ongoing)

### PR DESCRIPTION
**Pull request by adrianperreault**
Inspired by @icefoxen, this PR is an ongoing effort to improve documentation.

Ground to cover:
- [ ] command
    - [ ] mod
    - [ ] compute
    - [ ] graphics
    - [ ] raw
    - [ ] render_pass
    - [ ] transfer
- [ ] pso
    - [ ] mod
    - [ ] compute
    - [ ] descriptor
    - [ ] graphics
    - [ ] input_assembler
    - [ ] output_merger
- [ ] queue
    - [ ] mod
    - [ ] capability
    - [ ] family
    - [ ] submission
- [ ] adapter
- [ ] backend
- [ ] buffer
- [ ] device
- [ ] error
- [ ] format
- [ ] image
- [ ] lib
- [ ] mapping
- [ ] memory
- [ ] pass
- [ ] pool
- [ ] query
- [ ] queue
- [ ] range
- [ ] window

### CODE REVIEW

File: src/hal/src/image.rs [hunk 1]
```diff
@@ -589,7 +589,14 @@ pub struct SubresourceLayers {
     pub layers: Range<Layer>,
 }
 
-/// A subset of resources contained within an image.
+/// When creating an image view, used for describing the view's purpose and which part
```
Comment (msiglreith) [inline src/hal/src/image.rs]: I'm not sure if this is the appropiate documentation for `SubresourceRange`. This rather fits for `create_image_view`.

File: src/hal/src/image.rs [hunk 2]
```diff
@@ -589,7 +589,14 @@ pub struct SubresourceLayers {
...
+/// of the image should be accessed. Examples of purpose would be if the view should be
+/// used as a color, depth, or stencil target (or any combination). A range of mipmap levels
```
Comment (msiglreith) [inline src/hal/src/image.rs]: Yes, there is an aspect flag but it's not really indicating the usage purpose but rather with 'plane' of the image will be part of the subresource.

File: src/hal/src/pso/descriptor.rs [hunk 3]
```diff
@@ -1,4 +1,16 @@
 //! Descriptor sets and layouts.
+//!
+//! Resource descriptors are used by shaders for accessing resources like buffers
+//! and images. The use of a descriptor has three stages:
+//!
+//! - Descriptor layout is specified during pipeline creation
+//! - Descriptor set is allocated from a descriptor pool
+//! - Descriptor set is bound during rendering
```
Comment (msiglreith) [inline src/hal/src/pso/descriptor.rs]: Descriptor sets need also be updated (copy or write)

File: src/hal/src/pso/descriptor.rs [hunk 4]
```diff
@@ -1,4 +1,16 @@
...
+//!
+//! A Descriptor Set Layout describes the types of resources to be used by a pipeline.
+//!
+//! A Descriptor Set contains the actual resources to be bound/used: eg. buffer, images
```
Comment (msiglreith) [inline src/hal/src/pso/descriptor.rs]: Descriptor sets are actually views into resource, so an indirection. Basically a pointer to a memory slice and how it's supposed to be accessed by the shader.

File: src/hal/src/image.rs (hunk 2 above)
Comment (adrianperreault) [reply src/hal/src/image.rs]: Ok! The wording "purpose" came from the [Vulkan tutorial](https://vulkan-tutorial.com/Drawing_a_triangle/Presentation/Image_views). So the aspect flag is for 'filtering' which aspect of the image will be a part of the subresource?

File: src/hal/src/image.rs (hunk 1 above)
Comment (adrianperreault) [reply src/hal/src/image.rs]: Are SubresourceRanges used for anything other than describing details when creating image views?

**Comment by adrianperreault** [comment]
@icefoxen Sounds good! ☺️

**Comment by kvark** [comment]
Ping! Is this PR abandoned?

**Pull request closed without merging**